### PR TITLE
Clean up conda environment when `pip install` fails during `conda create`

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

conda doesn't clean up an environment when `pip install` fails during `conda create`. This behavior results in creating an incomplete conda environment and causes the following issue:

```bash
mlflow models serve ...
# fails a temporary network issue and leaves an incomplete conda environment

mlflow models serve ...
# serves the model in the incomplete environment created in the first attempt
```

## How is this patch tested?

Unit test

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/conda.py [hunk 1]
```diff
@@ -126,36 +131,55 @@ def get_or_create_conda_env(conda_env_path, env_id=None):
...
+                        "--channel",
+                        "conda-forge",
+                        "--override-channels",
+                        "-n",
+                        project_env_name,
+                        "python",
+                    ],
+                    stream_output=True,
+                )
+        except Exception:
```
Comment (WeichenXu123) [inline mlflow/utils/conda.py:166]: Even if subprocess succeeds, we need to also check CondaEnvException in subprocess stdout/stderr output.

Comment (WeichenXu123) [reply mlflow/utils/conda.py:166]: I suggest do:
```
exit_code, stdout, stderr = process.exec_cmd(..., throw_on_error=False)
if exit_code != 0 or "CondaEnvException" in stdout or "CondaEnvException" in stderr:
  # clean created env here...
```

File: mlflow/utils/conda.py [hunk 2]
```diff
@@ -126,36 +131,55 @@ def get_or_create_conda_env(conda_env_path, env_id=None):
...
+            if conda_env_path:
+                process.exec_cmd(
+                    [
+                        conda_env_create_path,
+                        "env",
+                        "create",
+                        "-n",
+                        project_env_name,
+                        "--file",
+                        conda_env_path,
```
Comment (WeichenXu123) [inline mlflow/utils/conda.py:148]: Question:
Shall we add `"--yes"` argument for `conda env create` command ?

File: mlflow/utils/conda.py (hunk 1 above)
Comment (harupy) [reply mlflow/utils/conda.py:166]: > Even if subprocess succeeds, we need to also check CondaEnvException in subprocess stdout/stderr output.

Why?

Comment (harupy) [reply mlflow/utils/conda.py:166]: I found pip failures causes conda to fail.

Comment (harupy) [reply mlflow/utils/conda.py:166]: Here's a proof:

```python
% cat conda.yaml
name: yes
channels:
  - conda-forge
dependencies:
  - python=3.7.12
  - pip
  - pip:
      - mlflow==999.999.999
                                                                                                                                                                 
% conda env create -f conda.yaml
Collecting package metadata (repodata.json): done
Solving environment: done
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Ran pip subprocess with arguments:
['/Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/yes/bin/python', '-m', 'pip', 'install', '-U', '-r', '/Users/harutakakawamura/Desktop/repositories/mlflow/condaenv.j5yolst2.requirements.txt']
Pip subprocess output:

Pip subprocess error:
ERROR: Could not find a version that satisfies the requirement mlflow==999.999.999 (from versions: 0.0.1, 0.1.0, 0.2.0, 0.2.1, 0.3.0, 0.4.0, 0.4.1, 0.4.2, 0.5.0, 0.5.1, 0.5.2, 0.6.0, 0.7.0, 0.8.0, 0.8.1, 0.8.2, 0.9.0, 0.9.0.1, 0.9.1, 1.0.0, 1.1.0, 1.1.1.dev0, 1.2.0, 1.3.0, 1.4.0, 1.5.0, 1.6.0, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.9.0, 1.9.1, 1.10.0, 1.11.0, 1.12.0, 1.12.1, 1.13, 1.13.1, 1.14.0, 1.14.1, 1.15.0, 1.16.0, 1.17.0, 1.18.0, 1.19.0, 1.20.0, 1.20.1, 1.20.2, 1.21.0, 1.22.0, 1.23.0, 1.23.1, 1.24.0)
ERROR: No matching distribution found for mlflow==999.999.999


CondaEnvException: Pip failed

                                                                                                                                                                 
% echo $?
1
```

Comment (harupy) [reply mlflow/utils/conda.py:166]: Please let me know if there are cases where conda succeeds when `CondaEnvException` is raised.

File: mlflow/utils/conda.py (hunk 2 above)
Comment (harupy) [reply mlflow/utils/conda.py:148]: Added `--yes` to `conda create` for safety.

File: mlflow/utils/conda.py (hunk 1 above)
Comment (BenWilson2) [reply mlflow/utils/conda.py:167]: I even tried installing an invalid python version (in this case, the conda env doesn't get built (obviously)) but it still returns (1) for this failure.

```
~/Documents via 🅒 base took 10s
➜ cat conda.yaml
name: hodor
channels:
  - conda-forge
dependencies:
  - python=3.14.1
  - pip
  - pip:
    - mlflow=42.42.42
(base)
~/Documents via 🅒 base
➜ conda env create -f conda.yaml
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound:
  - python=3.14.1

(base)
~/Documents via 🅒 base took 9s
➜ echo $?
1
➜ conda env list
# conda environments:
#
base                  *  /Users/benwilson/opt/anaconda3
diviner-dev              /Users/benwilson/opt/anaconda3/envs/diviner-dev
mlflow-dev-env           /Users/benwilson/opt/anaconda3/envs/mlflow-dev-env
```

Comment (harupy) [reply mlflow/utils/conda.py:167]: Thanks Ben for the investigation :)

Comment (harupy) [reply mlflow/utils/conda.py:167]: Given these results, I think the current approach is ok. @WeichenXu123 WDYT?

File: mlflow/utils/conda.py (hunk 2 above)
Comment (harupy) [reply mlflow/utils/conda.py:148]: Without `--yes`, `conda create` does ask for confirmation but it appears `subprocess` automatically accepts it (not 100% sure though).

Comment (BenWilson2) [reply mlflow/utils/conda.py:148]: Can confirm. 

``` bash
~/Documents via 🅒 base
➜ conda env list
# conda environments:
#
base                  *  /Users/benwilson/opt/anaconda3
diviner-dev              /Users/benwilson/opt/anaconda3/envs/diviner-dev
hodor                    /Users/benwilson/opt/anaconda3/envs/hodor
mlflow-dev-env           /Users/benwilson/opt/anaconda3/envs/mlflow-dev-env

(base)
~/Documents via 🅒 base took 2s
➜ (cat conda.yaml && conda env remove -n hodor)
name: hodor
channels:
  - conda-forge
dependencies:
  - python=3.9
  - pip
  - pip:
    - mlflow==1.20.0

Remove all packages in environment /Users/benwilson/opt/anaconda3/envs/hodor:

(base)
```
It subprocesses auto-accept :)

**Review by BenWilson2 — APPROVED** [review]
I think the implementation will work well :) LGTM!

**Pull request merged**